### DISCUSSION TITLE
How to derive a new state from different stores while rerending only when derived state changes?

### DISCUSSION THREAD

**Discussion by itsramiel** [pmndrs/zustand#2608] [Q&A] [answered]
Hello,

Sometimes I would like to derive a state from separated stores, and I would like to keep it as performant as possible, as in to only rerender when the derived state changes.

For example given the following two stores:

```
const usetCartStore = create<{
  items: Array<{ name: string; price: number }>;
}>()(() => ({
  items: [{ name: "t shirt", price: 5 }],
}));

const useCuoponsStore = create<{
  cuopons: Array<{ id: number; value: number }>;
}>()(() => ({
  cuopons: [
    { id: 1, value: 3 },
    { id: 2, value: 1.2 },
  ],
}));
```

I would like to get the following result:
```
  const cartTotal = usetCartStore((store) =>
    store.items.reduce((acc, item) => acc + item.price, 0),
  );
  const cuoponsTotal = useCuoponsStore((store) =>
    store.cuopons.reduce((acc, item) => acc + item.value, 0),
  );
  const total = cartTotal - cuoponsTotal;
```

However this would rerender the component when `cartTotal` or `cuoponsTotal` change but it makes sense that it only rerenders when `total` changes cause that's what I care about. Is there a way to do that?

I would imagine something similar to this:
```
  const total = useStoresSelectors(
    usetCartStore((store) =>
      store.items.reduce((acc, item) => acc + item.price, 0),
    ),
    (store) => store.cuopons.reduce((acc, item) => acc + item.value, 0),
    (cartTotal, cuoponsTotal) => cartTotal + cuoponsTotal,
  );
```

Is there such a thing?

**Comment by dbritto-dev** [comment]
@itsramiel this looks good to me:

```ts
const cartTotal = usetCartStore((store) =>
  store.items.reduce((acc, item) => acc + item.price, 0),
);
const cuoponsTotal = useCuoponsStore((store) =>
  store.cuopons.reduce((acc, item) => acc + item.value, 0),
);
const total = cartTotal - cuoponsTotal;
```

What you mean with this?

> However this would rerender the component when cartTotal or cuoponsTotal change but it makes sense that it only rerenders when total changes cause that's what I care about. Is there a way to do that?

For me sounds totally right, since total is derived from `cartTotal` and `cuoponsTotal` so makes sense that those cause re-renders

**Reply by itsramiel** [reply]
yes I know that the code I provided works, but it is not optimal. What I care about is `total` and not `cartTotal` and `cuoponsTotal`, individually. Assume that an action was done that changed both `carTotal` and `cuopinsTotal` but the `total` remained the same. The component shouldnt rerender because I only care about the `total` but it still rerenders in the code I provided.

Did I get my point through?

**Reply by dbritto-dev** [reply]
@itsramiel you should care about `cartTotal` and `couponsTotal` since `total` its derived from those. In the case you want to only care about `total` you could create memo(ized) component and pass total as props

**Reply by itsramiel** [reply]
Imaging the store was instead a single store:

```ts
const usetCartAndCuponsStore = create<{
  items: Array<{ name: string; price: number }>;
  cuopons: Array<{ id: number; value: number }>;
}>()(() => ({
  items: [{ name: "t shirt", price: 5 }],
  cuopons: [
    { id: 1, value: 3 },
    { id: 2, value: 1.2 },
  ],
}));
```

then I can do:

```ts
  const total = usetCartAndCuponsStore((store) => {
    return (
      store.items.reduce((acc, item) => acc + item.price, 0) +
      store.cuopons.reduce((acc, item) => acc + item.value, 0)
    );
  });
```

And this way the component only rerenders when `total` changes. What I want is this behavior but from two different stores and to know if this can be done internally in zustand without having to do manual memo myself.

**Reply by dbritto-dev** [reply, marked as answer]
@itsramiel well, zustand can't do that if you want something like that you could use `jotai` or `valtio`

**Reply by itsramiel** [reply]
alright thanks
